src/vm/amd64/pinvokestubs.S

```diff
@@ -0,0 +1,277 @@
+        
+        add             rsp, PInvokeStubForHost_STACK_FRAME_SIZE
+        ret
+NESTED_END PInvokeStubForHost, _TEXT
+
+
+PInvokeStubForHostInner_STACK_FRAME_SIZE = 0
+
+// integer registers saved in prologue
+PInvokeStubForHostInner_NUM_REG_PUSHES = 2
+PInvokeStubForHostInner_STACK_FRAME_SIZE = PInvokeStubForHostInner_STACK_FRAME_SIZE + PInvokeStubForHostInner_NUM_REG_PUSHES*8
+
+// Ensure that the new rsp will be 16-byte aligned.
+.if ((PInvokeStubForHostInner_STACK_FRAME_SIZE + 8) % 16) != 0
+PInvokeStubForHostInner_STACK_FRAME_SIZE = PInvokeStubForHostInner_STACK_FRAME_SIZE + 8
+.endif
+
+// Return address is immediately above the local variables.
+PInvokeStubForHostInner_RETURN_ADDRESS_OFFSET = PInvokeStubForHostInner_STACK_FRAME_SIZE
+PInvokeStubForHostInner_PARAM_REGISTERS_OFFSET = PInvokeStubForHostInner_RETURN_ADDRESS_OFFSET + 8
+
+PInvokeStubForHostInner_FRAME_OFFSET = PInvokeStubForHost_CALLEE_SCRATCH_SIZE
+
+// rdi - #stack args
+// rsi - PInvokeStubForHost's stack frame
+// rdx  - target address
+NESTED_ENTRY PInvokeStubForHostInner, _TEXT, NoHandler
+
+        push_nonvol_reg rbp
+        push_nonvol_reg r12
+        alloc_stack     PInvokeStubForHostInner_FRAME_OFFSET + PInvokeStubForHostInner_STACK_FRAME_SIZE - PInvokeStubForHostInner_NUM_REG_PUSHES*8
+        lea rbp, [rsp + PInvokeStubForHostInner_FRAME_OFFSET]
+        END_PROLOGUE
+
+        mov             r10, rdx
+        mov             r12, rsi
+
+        test            rdi, rdi
+        jnz             LOCAL_LABEL(HandleStackArgs)
+
+        //
+        // Allocate space for scratch area if there are no stack args.
+        //
+        sub             rsp, PInvokeStubForHost_CALLEE_SCRATCH_SIZE
+
+LOCAL_LABEL(DoneStackArgs):
+        // unspill args
+        mov             rdi, [r12 + PInvokeStubForHost_PARAM_REGISTERS_OFFSET +  0h]
+        mov             rsi, [r12 + PInvokeStubForHost_PARAM_REGISTERS_OFFSET +  8h]
+        mov             rdx,  [r12 + PInvokeStubForHost_PARAM_REGISTERS_OFFSET + 10h]
+        mov             rcx,  [r12 + PInvokeStubForHost_PARAM_REGISTERS_OFFSET + 18h]
+        movdqa          xmm0, [r12 + PInvokeStubForHost_XMM_SAVE_OFFSET +  0h]
+        movdqa          xmm1, [r12 + PInvokeStubForHost_XMM_SAVE_OFFSET + 10h]
+        movdqa          xmm2, [r12 + PInvokeStubForHost_XMM_SAVE_OFFSET + 20h]
+        movdqa          xmm3, [r12 + PInvokeStubForHost_XMM_SAVE_OFFSET + 30h]
+        movdqa          xmm4, [r12 + PInvokeStubForHost_XMM_SAVE_OFFSET + 40h]
+        movdqa          xmm5, [r12 + PInvokeStubForHost_XMM_SAVE_OFFSET + 50h]
+        movdqa          xmm6, [r12 + PInvokeStubForHost_XMM_SAVE_OFFSET + 60h]
+        movdqa          xmm7, [r12 + PInvokeStubForHost_XMM_SAVE_OFFSET + 70h]
+
+        call            r10
+        
+        // spill return value
+        mov             [r12 + PInvokeStubForHost_XMM_SAVE_OFFSET +  0h], rax                                                
```

Comment:
Please add a UNIXTODO: here to add spilling rdx and xmm1 when returning structs in registers on Unix is implemented.